docs/anti-patterns.mdx

```diff
@@ -4,35 +4,110 @@ title: Cadence Anti-Patterns
+### Problem
 
-If that resource has functions that emit events, any account can create an instance of that resource and emit those events.
+Some projects want to authenticate users using account addresses or want to perform operations for users directly.
+A not-uncommon reaction is to expect a user to pass their `AuthAccount` object as a function parameter
+to a contract function so that address can be queried or so an object can be stored directly.
+This is a problem, because the `AuthAccount` object can be used to access ANY private area of an account,
+meaning that a project could do something with your `AuthAccount` that you don't intend it to do.
+
+### Solution
+
+Projects should find other ways to authenticate users, such as using resources and capabilities as authentication objects.
+They should also expect to perform most storage and linking operations within transaction bodies
+instead of in a contract utility functions.
+
+There are some scenarios where using an `AuthAccount` object is necessary, such as a cold storage multi-sig,
+but those cases are extremely rare and `AuthAccount` usage should still be avoided unless absolutely necessary.
+
+## Auth References should be avoided
+
+### Problem
+
+[`auth` references](/language/references) should be avoided unless necessary:
+it allows downcasting any restricted type to its unrestricted type.
+The parent type could expose functionality that was not intended to be exposed.
+If the auth keyword is used on local variables they will be references.
+References are ephemeral and cannot be stored.
+This prevents any reference casting to be stored under account storage.
+However, if the auth keyword is used to store a public capability serious harm
+could be achieved since the value could be downcasted to a type that has functionality and values altered.
 
-If those events are meant to indicate actions taken using a single instance of that resource (for example an admin object, switchboard, or registry), or instances created through a particular workflow, then this will make event log search and management harder because events from other instances may be mixed in with the ones that you wish to examine.
+### Example
 
-To fix this, if there should be only a single instance of the resource it should be created and `link()`ed to a public path in an admin account's storage during the contracts's initializer.
+A common pattern in many NFT smart contracts is to include a public borrow function
+that borrows an auth reference to an NFT ([See Top Shot for an example](https://github.com/dapperlabs/nba-smart-contracts/blob/master/contracts/TopShot.cdc#L889-L906)).
+This allows anyone to access the stored metadata or extra fields that weren't part of the NFT standard.
+While this is fine in most scenarios, some NFTs might have privledged functions that shouldn't
+be exposed this way, so be careful when copying NFT projects that use this pattern. 
```

Comment:
A commonly seen pattern in NFT smart contracts is including a public borrow function that borrows an `auth` reference to an NFT (eg. [NBA Top Shot](https://github.com/dapperlabs/nba-smart-contracts/blob/master/contracts/TopShot.cdc#L889-L906)).  This allows anyone to access the stored metadata or extra fields that weren't part of the NFT standard. While generally safe in most scenarios, not all NFTs are built the same. Some NFTs may have admin functions that shouldn't be exposed by this method, so please be cautious & mindful when imitating NFT projects that use this pattern.